File: libs/collection/src/main/java/org/toolkit4j/collection/tree/TreeNode.java
package org.toolkit4j.collection.tree;

import java.util.Collection;
import java.util.stream.Stream;

/**
 * 树节点接口。TreeNode interface.
 *
 * <p>子类可细化 children 的集合类型，如 ListTreeNode、SetTreeNode。 Sub-interfaces may narrow children to List or
 * Set.
 *
 * @param <T> 节点数据类型
 */
public interface TreeNode<T> {

  /** 节点数据。Node data. */
  T data();

  /** 子节点集合。Children collection. 实现可为 List 或 Set 等。 */
  Collection<TreeNode<T>> children();

  /** 是否叶子节点。True if no children. */
  default boolean isLeaf() {
    return children().isEmpty();
  }

  /** 以该节点为根的 DFS 流。DFS stream of subtree. */
  default Stream<TreeNode<T>> stream() {
    return Stream.concat(Stream.of(this), children().stream().flatMap(TreeNode::stream));
  }
}

File: kit/collection/src/main/kotlin/org/toolkit4j/collection/tree/TreeNode.kt
// (C)2023
package org.toolkit4j.collection.tree

import java.util.Objects
import java.util.function.Consumer
import java.util.stream.Stream

interface TreeNode<T, out C : MutableCollection<T>> {
    var data: T

    var parent: TreeNode<T, @UnsafeVariance C>?

    val children: MutableCollection<TreeNode<T, @UnsafeVariance C>?>

    fun isRoot(): Boolean {
        return Objects.isNull(parent)
    }

    fun isLeaf(): Boolean {
        return children.isEmpty()
    }

    fun addChild(child: TreeNode<T, @UnsafeVariance C>?)

    fun addChildren(children: Collection<TreeNode<T, @UnsafeVariance C>?>)

    fun removeChild(child: TreeNode<T, @UnsafeVariance C>)

    fun stream(): Stream<TreeNode<T, @UnsafeVariance C>?>

    fun parallelStream(): Stream<TreeNode<T, @UnsafeVariance C>?>

    fun walk(action: Consumer<TreeNode<T, @UnsafeVariance C>?>)

    fun parallelWalk(action: Consumer<TreeNode<T, @UnsafeVariance C>?>)

    fun iterator(): Iterator<TreeNode<T, C>?>
}
